### ISSUE TITLE
Optimize replacement functionality

### ISSUE THREAD

**Issue by wetzelj** [pydicom/deid#117]
As discussed starting in [PR #112 (comment)](https://github.com/pydicom/deid/pull/112#issuecomment-594922571), optimize the processing of recipe actions to more efficiently perform defined actions.  

>I think we would want to parse an entire recipe first, build up a dependency graph of sorts (that also caches originally values that might be needed) and then do the replacements the most efficient way possible, possibly instead of parsing over every action and checking for every field, we first create an assignment of specific actions to specific fields, and then do a much more informed replacement.

**Comment by vsoch** [comment]
I think now that we have the DicomParser, we could look into discussing this again (but once you have some time @wetzelj, don't worry doesn't need to be soon!)

**Comment by vsoch** [comment]
Is this still something we want to discuss, perhaps in the near to distant future? :)

**Comment by wetzelj** [comment]
Yes, but not quite yet. We're a few weeks away from releasing our end-user application into user acceptance testing.  At that point, we'll see more testing with a diverse sets of images.  I suspect that we'll see some minor bugs with the header deidentification once this starts.  I'd like to get further along with this testing before starting on this optimization.
